crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -279,66 +295,343 @@ where
+                    self.ctx.read_bytes_post(key)?.map(Into::into)
+                else {
+                    // Ignore this event if it does not exist
+                    continue;
+                };
+                if packet_commitment == storage_commitment {
+                    seq_on_a = Some(path.sequence);
+                    break;
+                }
+            }
+        }
+        // If a key change with the same port, channel, and commitment as this
+        // message cannot be found, then ignore this message
+        if seq_on_a.is_none() {
+            return Ok(());
+        }
+
+        // Since IBC denominations are derived from Addresses
+        // when sending, we have to do a reverse look-up of the
+        // relevant token Address
+        let token = acc
+            .ibc_denoms
+            .get(&msg.message.packet_data.token.denom.to_string())
+            .cloned()
+            // If the reverse lookup failed, then guess the Address
+            // that might have yielded the IBC denom. However,
+            // guessing an IBC token address cannot possibly be
+            // correct due to the structure of query_ibc_denom
+            .or_else(|| {
+                Address::decode(msg.message.packet_data.token.denom.to_string())
+                    .ok()
+                    .filter(|x| {
+                        !matches!(
+                            x,
+                            Address::Internal(InternalAddress::IbcToken(_))
+                        )
+                    })
+            })
+            .ok_or_err_msg("Unable to decode IBC token")?;
+        let delta = ValueSum::from_pair(
+            token.clone(),
+            Amount::from_uint(Uint(*msg.message.packet_data.token.amount), 0)
+                .unwrap(),
+        );
+        // Required for the packet's receiver to get funds
+        let post_entry = acc
+            .post
+            .entry(ibc_taddr(msg.message.packet_data.receiver.to_string()))
+            .or_insert(ValueSum::zero());
+        // Enable funds to be received by the receiver in the
+        // IBC packet
+        *post_entry =
+            checked!(post_entry + &delta).map_err(native_vp::Error::new)?;
+
+        // If there is a transfer to the IBC account, then deduplicate the
+        // balance increase since we already accounted for it above
+        if is_sender_chain_source(
+            msg.message.port_id_on_a.clone(),
+            msg.message.chan_id_on_a.clone(),
+            &msg.message.packet_data.token.denom,
+        ) {
+            let post_entry =
+                acc.post.entry(addr_taddr(IBC)).or_insert(ValueSum::zero());
+            *post_entry =
+                checked!(post_entry - &delta).map_err(native_vp::Error::new)?;
+        }
+        Ok(())
+    }
+
+    // Apply the given write acknowledge to the changed balances structure
+    fn apply_write_ack(
+        &self,
+        acc: &mut ChangedBalances,
```

Comment:
Same comment as above (prefer returning a `ChangedBalances` to taking it as `mut`)